Implement alive across call for CSE candidates

briansull opened the pull request:
This is a Zero-Diff change
Using the alive across call information will take place in a future change.
This information will influence how aggressively we try to replace a CSE candidate into a new CSE LclVar.

AaronRobinsonMSFT commented on src/jit/compiler.h:
```diff
@@ -6088,8 +6088,9 @@ class Compiler
 
-    // Keeps tracked cse indices
-    BitVecTraits* cseTraits;
+    // trait informatiobn for CSE bitsets
```
~~informatiobn~~ -> information

AaronRobinsonMSFT commented on src/jit/morph.cpp:
```diff
@@ -8245,6 +8245,8 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
         }
     }
 
+    compCurBB->bbFlags |= BBF_HAS_CALL;
```
There are various places above where we return early. If indicating this here is okay, can you comment as to why the above doesn't count as a call?

briansull commented:
@dotnet/jit-contrib PTAL

This change updates the dataflow used by optCSE to properly compute the alive across call information.  It is store per CSE candidate in csdLiveAcrossCall 

As this is a Zero diff change we don't use this information with this change.  Instead the function LiveAcrossCall() will always return false  as it currently does today.

In a future change we can tune the CSE logic to be less aggressive when the CSE candidate is alive across a call.

mikedn commented on src/jit/optcse.cpp:
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
+
+                // Set the block->bbCseGen by or-ing in the afterCallGen set
+                //
+                BitVecOps::UnionD(cseTraits, block->bbCseGen, afterCallGen);
```
`afterCallGen` doesn't seem to be necessary - it collects CSE defs and then these defs get added to `bbCseGen`. So add the CSE defs directly to `bbCseGen`.

briansull replied:
OK I can add them directly to bbCseGen instead.

mikedn commented on src/jit/optcse.cpp:
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
+        // If the block contains a call we may have to update the bbCseGen
+        // as we may generate some CSEs after the last call in the block.
+        //
+        if ((block->bbFlags & BBF_HAS_CALL) != 0)
```
It would be better to make this `if (no call) continue`. It avoids additional indentation in the code below and makes it easy to see that the loop only looks at blocks with calls.

briansull replied:
OK

mikedn commented on src/jit/optcse.cpp:
```diff
@@ -882,37 +970,68 @@ class CSE_DataFlow
     CSE_DataFlow(Compiler* pCompiler) : m_pBitVecTraits(pCompiler->cseTraits), m_preMergeOut(BitVecOps::UninitVal())
     {
+#ifdef DEBUG
+        m_pCompiler = pCompiler;
```
Constructors should initialize data members in the initialization list.

briansull replied:
OK

mikedn commented on src/jit/optcse.cpp:
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
+            // This CSE is generated in 'block', we always set the cseBit
+            // If this block does not contain a call, we also set cseBit2
+            //
+            // If we have a call in this block the below we walk the trees backwards
```
"the below loop walks" or something like that I presume

mikedn commented on src/jit/optcse.cpp:
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
-            BitVecOps::AddElemD(cseTraits, block->bbCseGen, genCSEnum2bit(CSEindex));
+            BasicBlock* block   = lst->tslBlock;
+            unsigned    cseBit  = genCSEnum2bit(CSEindex) * 2;
+            unsigned    cseBit2 = cseBit + 1;
```
A more descriptive name would be useful. Perhaps `cseCrossCallBit`?

briansull replied:
The comment above states:

```
    // two bits are allocated per CSEcandidate,
    // the first one 'cseMaskTraits' tracks normal availability
    // the second one 'cseTraits' tracks availability when calls are considered killing
```

briansull replied:
I am open for suggestions on what names to use here.

CarolEidt replied:
I think that `cseCrossCallBit` seems reasonable, as suggested by @mikedn. 
That said, I find the comment that you reference above to be confusing. You say "two bits are allocated" and then say that the second one is `cseTraits`. But AFAICT, both bits you describe are in `cseTraits`.
I haven't fully reviewed this code yet, but it seems there are actually 3 bits per CSE candidate (2 in `cseTraits` and 1 in `cseMaskTraits`). You describe the latter as reflecting "normal availability", and I *think* the former has one bit for whether the cse is defined, and another for whether it is live across a call. If that's correct it would be good to make that clear in that comment above.

briansull replied:
You are right and that comment is incorrect.
These traits controls how many bits get allocated for a particular Bit Vector

There are two kinds needed now when dealing with CSEs:

`cseMaskTraits` only needs one bit per CSE candidate and it is only used by `optCSE_canSwap`
for `CSE_useMask` and `CSE_defMask` for a legality check.

I will rename the other one 'cseTraits' to 'cseLivenessTraits'. It needs two bits per CSE candidate plus one extra bit so that the liveness algorithm properly visits all children on the first visit.  

Note that I discovered this bug when I investigate couple of functions that had unexpected diffs. It previously existed, but was harder to hit because we over-allocated using BitVecTraits(EXPSET_SZ, this) on line 366, so could happen only when we have the maximum number of CSEs and an appropriate initial liveness state.

Now each CSE gets two bits, where a one bit means that the CSE is available and a 0 bit means that it is not available.   So at a definition you would see both bits changing from 0 to 1 and when visiting a call the second bit would get killed and become 0, while the other availability bit stays as 1.  I am currently calling the second bit `cseBit2`

I can call it `cseCrossCallBit` but it will go from 1 to 0 when it crosses a call.

briansull replied:
I renamed both bits: cseBit to cseAvailBit and cseBit2 to cseAvailCrossCallBit

mikedn commented on src/jit/optcse.cpp:
```diff
@@ -1239,6 +1403,54 @@ void Compiler::optValnumCSE_Availablity()
+                        {
+
+                            EXPSET_TP call_killset = BitVecOps::UninitVal();
+                            BitVecOps::Assign(cseTraits, call_killset, call_kills_all_cses);
```
This will always allocate a bitset when the number of CSEs exceeds 32. You could hoist this bitset out of the loop to avoid this. But this acrobatics seems unnecessary, you could just save `cseBit2` bit from `available_cses` to a variable and restore it after doing the intersection.

briansull commented:
@dotnet/jit-contrib PTAL

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

briansull commented:
@CarolEidt PTAL

CarolEidt commented on src/jit/compiler.h:
```diff
@@ -6115,8 +6115,9 @@ class Compiler
-    BitVecTraits* cseTraits;
+    // trait information for CSE bitsets
+    BitVecTraits* cseMaskTraits;     // one bit per CSE candidate
+    BitVecTraits* cseLivenessTraits; // two bits per CSE candidate + 1
```
You should add a comment here (similar to your response to my question before) about what these are used for.

CarolEidt commented on src/jit/optcse.cpp:
```diff
@@ -788,9 +798,22 @@ void Compiler::optCseUpdateCheckedBoundMap(GenTree* compare)
+    // Init traits and full/empty bitvectors.  This will be used to track the
+    // individual cse indexes.
+
+    // two bits, plus one extra bit is allocated per CSEcandidate to compute CSE liveness
```
I would reword this as:
```
// We allocate two bits per CSE candidate, plus an extra bit to handle the initial unvisited case.
// The first bit tracks ...
```
And somewhere it would be good to describe how that extra bit is used to handle the unvisited case.

CarolEidt commented on src/jit/optcse.cpp:
```diff
@@ -1013,23 +1187,47 @@ void Compiler::optValnumCSE_Availablity()
+                        madeLiveAcrossCall      = true;
+                    }
+
+                    // If this a a CSE def, the the call-kill bit should also be a def (i.e not set in the
```
typo "the the" should be "then the" I think.
It is confusing to say that "the call-kill bit should also be a def". I think it would be clearer to change this comment to:
```
// If this is a CSE def (i.e. the CSE is not available here, since it is being defined), then the call-kill bit
// should also be zero since it is also not available across a call.
```

CarolEidt commented on src/jit/optcse.cpp:
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
+        : m_comp(pCompiler), m_preMergeOut(BitVecOps::UninitVal()), m_callKills(BitVecOps::UninitVal())
     {
+        m_callKills = BitVecOps::MakeEmpty(m_comp->cseLivenessTraits);
+        for (unsigned inx = 0; inx < m_comp->optCSECandidateCount; inx++)
```
It seems that you should be able to factor out this code so that it isn't duplicated below.

CarolEidt commented on src/jit/optcse.cpp:
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
+            unsigned CseAvailBit = inx * 2;
+
+            // a one preserves availability and a zero kills the availability
+            // we generate this kind of bit pattern:  101010101010
```
If I understand correctly, these two bits really reflect 3 states:
- The CSE is available and is not killed by a call - `11`
- The CSE is available, but is killed by a call - `10`
- The CSE is not available - `00`

The `10` pattern seems counter-intuitive as an initial setting. If you want to initialize it to "full" it would seem logical to turn on both bits, and if you want to initialize it to "empty" it would seem logical to turn off both bits. What am I missing here?

briansull replied:
Yes, your understanding is basically correct, but I would express it this way:

`11` - The CSE is available, and is also available when considering calls as killing availability.
`10 `- The CSE is available, but is not available when considering calls as killing availability.
`00`  - The CSE is not available
`01 `- Illegal combination

This code constructs the mask `m_callKills` which is used below to clear the lower bit, - `10` of the `bbCseIn` for every CSE candidate, when the block contains a callsite:

```
            if (block->bbFlags & BBF_HAS_CALL)
            {
                BitVecOps::IntersectionD(m_comp->cseLivenessTraits, cseIn_withCallsKill, m_callKills);
            }
```

briansull replied:
I can rename `m_callKills` to `m_callKillsMask` to make this more clear.

briansull commented on src/jit/optcse.cpp:
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
     void StartMerge(BasicBlock* block)
     {
-        BitVecOps::Assign(m_pBitVecTraits, m_preMergeOut, block->bbCseOut);
+        BitVecOps::Assign(m_comp->cseLivenessTraits, m_preMergeOut, block->bbCseOut);
```
This initializes `m_preMergeOut` to `cseFull`, which is all bits set,  because we previous set `bbCseOut` to `cseFull` in `optValnumCSE_InitDataFlow()`

```
cseFull           = BitVecOps::MakeFull(cseLivenessTraits);
block->bbCseOut = BitVecOps::MakeCopy(cseLivenessTraits, cseFull);
```

briansull commented on src/jit/optcse.cpp:
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
+        }
+        BitVecOps::DataFlowD(m_comp->cseLivenessTraits, block->bbCseOut, block->bbCseGen, cseIn_withCallsKill);
+
+        bool result = !BitVecOps::Equal(m_comp->cseLivenessTraits, block->bbCseOut, m_preMergeOut);
```
This is why we need the extra bit:

This dataflow algorithm will push all of this block successors onto the `worklist `to visit when this `EndMerge` step returns true:

In dataflow.h line 63:
```
        if (callback.EndMerge(block))
        {
            for (BasicBlock* succ : block->GetAllSuccs(m_pCompiler))
            {
                worklist.insert(worklist.end(), succ);
            }
        }
```

If the very first time that we vist this block we compute that `block->bbCseOut`, has all bits set, (i.e `cseFull`)
Then we won't visit this blocks successors at all.

Before my changes here, we always used `unsigned (-1) ` as the value for cseFull, this meant that we essentially always had one or more extra bits set in `bbCseOut`.  

I found this issue after my change because there were a couple of methods that had blocks that made every CSE available and then the algorithm didn't properly visit successor blocks.
`

briansull commented:
@dotnet/jit-contrib  PTAL - I have addressed all of the feedback from Carol and have tested this again, with zero diffs.  It will enable better CSE choices in a future check in.
